pkg/util/retry/retry.go

```diff
@@ -116,53 +145,138 @@ func (r *Retry) Next() bool {
+	)
+
+	if r.backingOffHook != nil {
+		r.backingOffHook(backoff)
 	}
+
 	select {
-	case <-time.After(d):
-		r.currentAttempt++
-		return true
 	case <-r.opts.Closer:
 		return false
 	case <-r.ctx.Done():
 		return false
+	case <-backoffCh:
+		r.currentAttempt++
+		return true
+	case <-maxDurExpCh:
+		return false
 	}
 }
 
-// closedC is returned from Retry.NextCh whenever a retry
-// can begin immediately.
-var closedC = func() chan time.Time {
-	c := make(chan time.Time)
+func (r *Retry) retryLimitReached() bool {
+	return (r.opts.MaxRetries > 0 && r.currentAttempt >= r.opts.MaxRetries) ||
+		(r.opts.MaxDuration > 0 && !r.opts.Clock.Now().Before(r.deadline))
+}
+
+// immediateCh creates a channel that is immediately written to with the
+// provided value and then closed.
+func immediateCh(v bool) chan bool {
+	c := make(chan bool, 1)
+	c <- v
 	close(c)
 	return c
-}()
+}
+
+// timerCh returns two channels: one for the backoff timer and one for the
+// maximum duration expiration timer. It also returns a cleanup function that
+// stops both timers. A time.Time value is sent on the channels when their
+// respective timers expire.
+func (r *Retry) timerCh(
+	backoff time.Duration,
+) (backoffCh, maxDurExpCh <-chan time.Time, cleanup func()) {
+	var maxDurTimer timeutil.TimerI
+	if r.opts.MaxDuration > 0 {
+		maxDurTimer = r.opts.Clock.NewTimer()
+		maxDurTimer.Reset(max(r.opts.Clock.Until(r.deadline), 0))
+		maxDurExpCh = maxDurTimer.Ch()
+	}
+
+	backoffTimer := r.opts.Clock.NewTimer()
+	backoffTimer.Reset(backoff)
+
+	cleanup = func() {
+		if maxDurTimer != nil {
+			maxDurTimer.Stop()
+		}
+		backoffTimer.Stop()
+	}
+	return backoffTimer.Ch(), maxDurExpCh, cleanup
+}
 
 // NextCh returns a channel which will receive when the next retry
-// interval has expired.
-func (r *Retry) NextCh() <-chan time.Time {
+// interval has expired. If the received value is true, it indicates a retry
+// should be made. If the received value is false, it indicates that no retry
+// should be made.
+// Note: This does not respect the Closer or context cancellation and it is the
+// caller's responsibility to manage the lifecycle.
+func (r *Retry) NextCh() <-chan bool {
 	if r.isReset {
 		r.isReset = false
-		return closedC
+		return immediateCh(true)
+	}
+
+	backoff := r.retryIn()
+	if r.retryLimitReached() || r.shouldPreemptivelyCancel(backoff) {
+		return immediateCh(false)
 	}
-	r.currentAttempt++
-	if r.opts.MaxRetries > 0 && r.currentAttempt > r.opts.MaxRetries {
-		return nil
+
+	backoffCh, maxDurExpCh, cleanup := r.timerCh(backoff)
+	if r.backingOffHook != nil {
+		r.backingOffHook(backoff)
 	}
-	return time.After(r.retryIn())
+
+	log.VEventfDepth(
+		r.ctx, 1 /* depth */, 2 /* level */, "will retry after %s", backoff,
+	)
+
+	ch := make(chan bool, 1)
+	go func() {
+		defer cleanup()
+		defer close(ch)
+		select {
+		case <-backoffCh:
+			r.currentAttempt++
+			ch <- true
+		case <-maxDurExpCh:
+			ch <- false
+		}
+	}()
```

Comment:
One note is that even if you aren't going to pre-emptively cancel you know ahead of time which timer should trigger first.  I wonder if this could end up looking like:

```golang
	ch := make(chan bool)
	duration, shouldAttempt := r.calcBackoff()
	if shouldAttempt {
		r.currentAttempt++
	}
	time.AfterFunc(duration, func() {
		defer close(ch)
		ch <- shouldAttempt
	})
	return ch
```